core/domain/stats_services.py

```diff
@@ -212,6 +213,165 @@ def handle_stats_creation_for_new_exp_version(
     create_stats_model(exploration_stats)
 
 
+def create_exp_issues_for_new_exploration(exp_id, exp_version):
+    """Creates the ExplorationIssuesModel instance for the exploration.
+
+    Args:
+        exp_id: str. ID of the exploration.
+        exp_version: int. Version of the exploration.
+    """
+    stats_models.ExplorationIssuesModel.create(exp_id, exp_version, [])
+
+
+def update_exp_issues_for_new_exp_version(exploration, change_list):
+    """Retrieves the ExplorationIssuesModel for the old exp_version and makes
+    any required changes to the structure of the model.
+
+    Args:
+        exploration: Exploration. Domain object for the exploration.
+        change_list: list(dict). A list of changes introduced in this commit.
+    """
+    # TODO(pranavsid98): Convert below two lines to use get_by_id() instead.
+    exp_issues_model = stats_models.ExplorationIssuesModel.get_model(
+        exploration.id, exploration.version - 1)
+    if exp_issues_model is None:
+        create_exp_issues_for_new_exploration(
+            exploration.id, exploration.version - 1)
+        return
+    exp_issues = get_exp_issues_from_model(exp_issues_model)
+
+    # We make a mapping from state name to list of exploration issues so that
```

Comment:
I'm not sure I understand why we're doing it this way -- it seems a little indirect, and hard to extend to other types of issues in the future.

Why not do a for loop on the issues, and on each issue, check whether any of its state names are affected? That seems simpler to understand and less prone to error.

Reply:
@seanlip I'll do a raw analysis here comparing approach 1 (my current approach), and approach 2 (your suggested approach).

Assume:
I = num_issues
DS = num_deleted_states
RS = num_renamed_states
P = num_playthroughs
A = num_actions

Approach 1 worst case time complexity:

I + (DS*I) + (RS*(I+(P*A))) 
= (I * (1 + DS + RS)) + RS*P*A 

Approach 2  worst case time complexity: 

I * (DS + (RS *P*A))) 

Now, making some core assumptions which usually hold:
- I, P, A considerably add more weight than DS, RS.
- I, P, A can be reasonably estimated to A each in worst case, as A will be the higher value in usual cases.
- DS, RS can be reasonably reduced to 1 in comparison to I, P, A in larger versions of the model

Equation 1 can be reduced to: (A*3) + A*A = O(A squared)
Equation 2 can be reduced to: A * (1 + (A*A)) = A + A*A*A = O(A cubed)

What do you think about this?

Reply:
This is a bit of an odd analysis. For example, you can't really say I ~ A "because A is the higher value in most cases". That introduces a dependence between I and A. Same thing for P.

A better way to frame this might be: I = num_issues, P = mean_num_playthroughs_per_issue, A = mean_num_actions_per_playthrough, DS = num_deleted_states_per_commit, RS = num_renamed_states_per_commit. That way, each of the variables are independent.

Try again?

Reply:
Also, here is my reasoning for the suggestion, tell me if you think I'm going wrong anywhere:

- We will have to look at every action at some point anyway in this analysis. That's already O(IPA) (using the variables I defined above) in order to know whether it's affected.
- If we do some preprocessing by using the version diff to build a dict that maps state names to instructions on what to do with them (either delete, or rename to X), then updating each action would be O(1).
- So overall the complexity is O(IPA) and you can't go any lower.

Reply:
Btw, this is kind of an expensive operation. Are you planning to do it asynchronously so that the exploration editor doesn't need to wait? (You can do this by passing an event into the deferred taskqueue or similar.)

Also -- if you are keeping invalid issues around then how are you going to handle this step on those issues?

Reply:
@seanlip 

O(IPA) - does it need to hold always?
So, what if I run a first pass through the issues, and append all playthrough_ids to a list and retrieve them. I can reduce it to O(I+PA), no?

Re deferred task queue, yes, I think we should do that.
Re invalid issues, refer to comment above.

Reply:
If you look at the definition I had of P (playthroughs _per issue_), you'll see that it's still O(IPA). Every playthrough is still being considered once.